velox/common/memory/SharedArbitrator.cpp

```diff
@@ -419,12 +455,39 @@ uint64_t SharedArbitrator::decrementFreeCapacityLocked(
   return allocatedBytes;
 }
 
+uint64_t SharedArbitrator::getCapacityShrinkTarget(
+    const MemoryPool& pool,
+    uint64_t requestBytes) const {
+  const auto capacity = pool.capacity();
+  if (pool.reservedBytes() == 0 && requestBytes == capacity) {
+    // When the pool is at the end of its lifetime, this always happens. We
+    // don't change the target in this case to allow the full release of the
+    // pool's capacity. There might be other corner cases when the pool has not
+    // finished and gets into this if condition. It is okay to not respect min
+    // free capacity in those cases as they are rare.
+    return requestBytes;
+  }
+  auto targetBytes = requestBytes;
+  if (memoryPoolMinFreeCapacity_ != 0) {
+    const auto minFreeBytes = std::min(
+        static_cast<uint64_t>(capacity * memoryPoolMinFreeCapacityPct_),
+        memoryPoolMinFreeCapacity_);
+    const auto maxShrinkBytes = std::max(
+        static_cast<int64_t>(0),
+        static_cast<int64_t>(
+            pool.freeBytes() - static_cast<int64_t>(minFreeBytes)));
```

Comment:
Are all these conversions are required?

Reply:
Yeah they are. static_cast<int64_t>(minFreeBytes) makes sure the subtraction does not go wild as both are uint64_t type. the outer cast is to make std::max() happy across compilers.